File: libraries/python/getpatter/llm/hermes.py
"""Hermes agent-runtime LLM for Patter pipeline mode.

Thin preset over :class:`getpatter.llm.openai_compatible.OpenAICompatibleLLMProvider`
that defaults the base URL, model, env-key name, and timeout for the Hermes
agent runtime. Patter is the voice shell (carrier + STT + turn-taking + TTS);
Hermes is the brain on the line — each turn is one
``POST {base_url}/chat/completions`` against the local Hermes gateway.

Hermes keys sessions off the OpenAI ``user`` field, so the per-call
``session_user_prefix`` defaults to ``"patter-call-"`` — one phone call maps to
one Hermes session.
"""

from __future__ import annotations

import os
from typing import ClassVar

from getpatter.llm.openai_compatible import OpenAICompatibleLLMProvider

__all__ = ["LLM"]

# Hermes gateway default (loopback; operator-co-located deployment).
_BASE_URL = "http://127.0.0.1:8642/v1"
_DEFAULT_MODEL = "hermes-agent"


class LLM(OpenAICompatibleLLMProvider):
    """Hermes agent-runtime LLM provider.

    Example::

        from getpatter.llm import hermes

        llm = hermes.LLM()                         # all env-defaulted
        llm = hermes.LLM(model="hermes-7b")        # explicit model override
        llm = hermes.LLM(api_key="...", base_url="http://host:8642/v1")

    Defaults:

    * ``base_url`` → ``http://127.0.0.1:8642/v1``
    * ``model`` → ``API_SERVER_MODEL_NAME`` env if set, else ``"hermes-agent"``
    * ``api_key`` → ``api_key`` arg or ``API_SERVER_KEY`` env (may be absent for
      a keyless local Hermes)
    * ``timeout`` → ``120.0`` s (runtimes run tools / memory / skills before
      replying, so turns can take 30-90 s)
    * ``session_user_prefix`` → ``"patter-call-"`` (Hermes keys sessions off
      the OpenAI ``user`` field)
    """

    provider_key: ClassVar[str] = "hermes"

    def __init__(
        self,
        api_key: str | None = None,
        *,
        base_url: str = _BASE_URL,
        model: str | None = None,
        timeout: float = 120.0,
        **kwargs,
    ) -> None:
        resolved_model = model or os.environ.get(
            "API_SERVER_MODEL_NAME", _DEFAULT_MODEL
        )
        super().__init__(
            api_key=api_key,
            base_url=base_url,
            model=resolved_model,
            api_key_env="API_SERVER_KEY",
            timeout=timeout,
            session_user_prefix="patter-call-",
            **kwargs,
        )

File: libraries/typescript/src/llm/hermes.ts
/**
 * Hermes agent-runtime LLM preset for Patter's pipeline mode.
 *
 * Thin preset over {@link OpenAICompatibleLLMProvider}: defaults the base URL,
 * model, env-key name, timeout, and session-continuity prefix for the Hermes
 * agent runtime so a user just writes ``phone.agent({ llm: new hermes.LLM() })``.
 *
 * Hermes runs tools / memory / skills internally before replying, so a single
 * conversation turn can take 30-90 s — hence the 120 s default timeout. It
 * keys sessions off the OpenAI ``user`` field, so the preset enables the
 * ``patter-call-`` session prefix to give one runtime session per phone call.
 */
import {
  OpenAICompatibleLLMProvider,
  type OpenAICompatibleLLMOptions,
} from './openai-compatible';

/** Default Hermes agent-runtime base URL (loopback, operator-controlled). */
const BASE_URL = 'http://127.0.0.1:8642/v1';
/** Fallback model when neither ``model`` nor ``API_SERVER_MODEL_NAME`` is set. */
const DEFAULT_MODEL = 'hermes-agent';
/** Env var Hermes reads its bearer from. */
const API_KEY_ENV = 'API_SERVER_KEY';
/** Env var Hermes reads its model id from. */
const MODEL_ENV = 'API_SERVER_MODEL_NAME';
/** Per-call session prefix → one Hermes session per phone call. */
const SESSION_USER_PREFIX = 'patter-call-';
/** Default timeout (seconds): runtimes run tools before replying. */
const DEFAULT_TIMEOUT_S = 120;

/** Constructor options for the Hermes ``LLM`` preset. */
export interface HermesLLMOptions {
  /** Bearer token. Falls back to ``API_SERVER_KEY`` env var when omitted. */
  apiKey?: string;
  /** Override the Hermes base URL (rarely needed). */
  baseUrl?: string;
  /** Model id. Falls back to ``API_SERVER_MODEL_NAME`` env, then ``"hermes-agent"``. */
  model?: string;
  /** Per-request timeout in seconds. Default ``120``. */
  timeout?: number;
  /** Extra headers merged after the SDK ``User-Agent``. */
  extraHeaders?: Record<string, string>;
  /** Sampling temperature [0, 2]. */
  temperature?: number;
  /** Max tokens in the assistant response (sent as ``max_completion_tokens``). */
  maxTokens?: number;
  /** OpenAI-style ``response_format`` for JSON mode / structured outputs. */
  responseFormat?: Record<string, unknown>;
  /** Whether to allow parallel tool calls. */
  parallelToolCalls?: boolean;
  /** ``"auto" | "none" | "required"`` or a specific tool object. */
  toolChoice?: string | Record<string, unknown>;
  /** Sampling seed for reproducible outputs. */
  seed?: number;
  /** Nucleus sampling cutoff in [0, 1]. */
  topP?: number;
  /** Penalty in [-2, 2] applied to repeated tokens. */
  frequencyPenalty?: number;
  /** Penalty in [-2, 2] applied to seen tokens. */
  presencePenalty?: number;
  /** Stop sequence(s). */
  stop?: string | string[];
}

/**
 * Hermes agent-runtime LLM provider (OpenAI-compatible, streaming).
 *
 * @example
 * ```ts
 * import * as hermes from "getpatter/llm/hermes";
 * const llm = new hermes.LLM();                       // env-defaulted, keyless OK
 * const llm = new hermes.LLM({ apiKey: "...", model: "hermes-7b" });
 * ```
 */
export class LLM extends OpenAICompatibleLLMProvider {
  static readonly providerKey = 'hermes';

  constructor(opts: HermesLLMOptions = {}) {
    const model = opts.model ?? process.env[MODEL_ENV] ?? DEFAULT_MODEL;
    const options: OpenAICompatibleLLMOptions = {
      apiKey: opts.apiKey,
      apiKeyEnv: API_KEY_ENV,
      baseUrl: opts.baseUrl ?? BASE_URL,
      model,
      timeout: opts.timeout ?? DEFAULT_TIMEOUT_S,
      sessionUserPrefix: SESSION_USER_PREFIX,
      extraHeaders: opts.extraHeaders,
      temperature: opts.temperature,
      maxTokens: opts.maxTokens,
      responseFormat: opts.responseFormat,
      parallelToolCalls: opts.parallelToolCalls,
      toolChoice: opts.toolChoice,
      seed: opts.seed,
      topP: opts.topP,
      frequencyPenalty: opts.frequencyPenalty,
      presencePenalty: opts.presencePenalty,
      stop: opts.stop,
    };
    super(options);
  }
}
